### PR TITLE
Avg and max velocity

### PR DESCRIPTION
**Pull request by rtldg**
For #646. ~~Adds `native void Shavit_GetAvgMaxVelocity(int client, float &avg, float &max)`.~~
Adds `Shavit_GetAvgVelocity()`, `Shavit_GetMaxVelocity()`, `Shavit_SetAvgVelocity()`, `Shavit_SetMaxVelocity()`.
Adds `float avgvel, float maxvel` to `Shavit_OnWorldRecord`, `Shavit_OnFinish_Post`, `Shavit_OnFinish`, `Shavit_OnWorstRecord`.
Adds avg/max velocity to `cp_cache_t` and `timer_snapshot_t`
Prints avg/max velocity to run-finish message.
The avg/max velocity should probably be stored in the db eventually too... but we'll wait for that

### CODE REVIEW

**Comment by GAMMACASE** [comment]
Isn't it better to have 2 different natives tho? Like currently I would need to create dummy variable to get avg or max speed only.

**Comment by kidfearless** [comment]
Why should this be added directly to the timer instead, instead of with a third-party plugin?

**Comment by Nairdaa** [comment]
> Why should this be added directly to the timer instead, instead of with a third-party plugin?

Because it should be a part of the timer and not a third-party plugin.

**Comment by kidfearless** [comment]
> 
> 
> > Why should this be added directly to the timer instead, instead of with a third-party plugin?
> 
> Because it should be a part of the timer and not a third-party plugin.

Because it should isn't a very good reason to add something to the timer that serves no purpose to the timer.

**Comment by Nairdaa** [comment]
> > > Why should this be added directly to the timer instead, instead of with a third-party plugin?
> > 
> > 
> > Because it should be a part of the timer and not a third-party plugin.
> 
> Because it should isn't a very good reason to add something to the timer that serves no purpose to the timer.

It's been a thing in timers since I started to bhop back in like 2012-2013. It's just another useless stat to see, but cool to have, just like the amount of jumps, strafes or sync in the hud.

File: addons/sourcemod/scripting/include/shavit.inc [hunk 1]
```diff
@@ -1840,6 +1840,18 @@ native int Shavit_PrintToChat(int client, const char[] format, any ...);
...
+ * Gets the average and max velocity of a player's jumps.
+ * Velocity is calculated as a 2d vector.
+ * Average calculation: avg += (vel - avg) / jumps
+ *
+ * @param client					Client index
+ * @param avg						Variable to recieve the average velocity
+ * @param max						Variable to receive the max velocity
+ * @noreturn
+ */
+native void Shavit_GetAvgMaxVelocity(int client, float &avg, float &max);
```
Comment (kidfearless) [inline addons/sourcemod/scripting/include/shavit.inc:1853]: Create separate getters and setters for each variable

File: addons/sourcemod/scripting/shavit-core.sp [hunk 2]
```diff
@@ -1436,6 +1439,15 @@ void VelocityChanges(int data)
...
 	}
+
+	if (GetTimerStatus(client) == view_as<int>(Timer_Running))
+	{
+		float curVel = SquareRoot(Pow(fAbsVelocity[0], 2.0) + Pow(fAbsVelocity[1], 2.0));
+		float maxVel = gA_Timers[client].fMaxVelocity;
+		gA_Timers[client].fMaxVelocity = (curVel > maxVel) ? curVel : maxVel;
+		// STOLEN from Epic/Disrevoid. Thx :)
+		gA_Timers[client].fAvgVelocity += (curVel - gA_Timers[client].fAvgVelocity) / gA_Timers[client].iJumps;
+	}
```
Comment (kidfearless) [inline addons/sourcemod/scripting/shavit-core.sp:1450]: Implement the average as a sum / total frames. That way styles that don't jump can still benefit from an average velocity counter.

**Comment by Nairdaa** [comment]
Did you ever split this into two messages for scroll styles?

**Comment by rtldg** [comment]
yeah it's two messages right now but I never dumped the avg/max velo to the db

**Pull request merged**